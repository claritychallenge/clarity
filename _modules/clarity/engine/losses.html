

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../../">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>clarity.engine.losses &mdash; pyClarity 0.7.1.post21 documentation</title>
      <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="../../../_static/css/theme.css?v=e59714d7" />
      <link rel="stylesheet" type="text/css" href="../../../_static/custom.css" />

  
      <script src="../../../_static/jquery.js?v=5d32c60e"></script>
      <script src="../../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../../../_static/documentation_options.js?v=5cc28f93"></script>
      <script src="../../../_static/doctools.js?v=9a2dae69"></script>
      <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
      <script type="module" src="https://cdn.jsdelivr.net/npm/mermaid@11.2.0/dist/mermaid.esm.min.mjs"></script>
      <script type="module" src="https://cdn.jsdelivr.net/npm/@mermaid-js/layout-elk@0.1.4/dist/mermaid-layout-elk.esm.min.mjs"></script>
      <script type="module">import mermaid from "https://cdn.jsdelivr.net/npm/mermaid@11.2.0/dist/mermaid.esm.min.mjs";import elkLayouts from "https://cdn.jsdelivr.net/npm/@mermaid-js/layout-elk@0.1.4/dist/mermaid-layout-elk.esm.min.mjs";mermaid.registerLayoutLoaders(elkLayouts);mermaid.initialize({startOnLoad:false});</script>
      <script src="https://cdn.jsdelivr.net/npm/d3@7.9.0/dist/d3.min.js"></script>
      <script type="module">
import mermaid from "https://cdn.jsdelivr.net/npm/mermaid@11.2.0/dist/mermaid.esm.min.mjs";
window.addEventListener("load", () => mermaid.run());
</script>
    <script src="../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../index.html" class="icon icon-home">
            pyClarity
              <img src="../../../_static/challenges.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../api.html">API Reference</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">pyClarity</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../index.html">Module code</a></li>
      <li class="breadcrumb-item active">clarity.engine.losses</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for clarity.engine.losses</h1><div class="highlight"><pre>
<span></span><span class="kn">import</span><span class="w"> </span><span class="nn">torch</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">clarity.predictor.torch_stoi</span><span class="w"> </span><span class="kn">import</span> <span class="n">NegSTOILoss</span>


<div class="viewcode-block" id="SISNRLoss">
<a class="viewcode-back" href="../../../clarity.engine.html#clarity.engine.losses.SISNRLoss">[docs]</a>
<span class="k">class</span><span class="w"> </span><span class="nc">SISNRLoss</span><span class="p">(</span><span class="n">torch</span><span class="o">.</span><span class="n">nn</span><span class="o">.</span><span class="n">Module</span><span class="p">):</span>
    <span class="c1"># def __init__(self):  # removed as does nothing</span>
    <span class="c1">#    super(SISNRLoss, self).__init__()</span>

<div class="viewcode-block" id="SISNRLoss.cal_sisnr">
<a class="viewcode-back" href="../../../clarity.engine.html#clarity.engine.losses.SISNRLoss.cal_sisnr">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">cal_sisnr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">x</span><span class="p">,</span> <span class="n">s</span><span class="p">,</span> <span class="n">eps</span><span class="o">=</span><span class="mf">1e-8</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Arguments:</span>
<span class="sd">        x: separated signal, N x S tensor</span>
<span class="sd">        s: reference signal, N x S tensor</span>
<span class="sd">        Return:</span>
<span class="sd">        sisnr: N tensor</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">def</span><span class="w"> </span><span class="nf">l2norm</span><span class="p">(</span><span class="n">mat</span><span class="p">,</span> <span class="n">keepdim</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
            <span class="k">return</span> <span class="n">torch</span><span class="o">.</span><span class="n">norm</span><span class="p">(</span><span class="n">mat</span><span class="p">,</span> <span class="n">dim</span><span class="o">=-</span><span class="mi">1</span><span class="p">,</span> <span class="n">keepdim</span><span class="o">=</span><span class="n">keepdim</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">x</span><span class="o">.</span><span class="n">shape</span> <span class="o">!=</span> <span class="n">s</span><span class="o">.</span><span class="n">shape</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Dimension mismatch when calculate si-snr, </span><span class="si">{</span><span class="n">x</span><span class="o">.</span><span class="n">shape</span><span class="si">}</span><span class="s2"> vs </span><span class="si">{</span><span class="n">s</span><span class="o">.</span><span class="n">shape</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="p">)</span>
        <span class="n">x_zm</span> <span class="o">=</span> <span class="n">x</span> <span class="o">-</span> <span class="n">torch</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">dim</span><span class="o">=-</span><span class="mi">1</span><span class="p">,</span> <span class="n">keepdim</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">s_zm</span> <span class="o">=</span> <span class="n">s</span> <span class="o">-</span> <span class="n">torch</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">s</span><span class="p">,</span> <span class="n">dim</span><span class="o">=-</span><span class="mi">1</span><span class="p">,</span> <span class="n">keepdim</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">t</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">torch</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">x_zm</span> <span class="o">*</span> <span class="n">s_zm</span><span class="p">,</span> <span class="n">dim</span><span class="o">=-</span><span class="mi">1</span><span class="p">,</span> <span class="n">keepdim</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="o">*</span> <span class="n">s_zm</span>
            <span class="o">/</span> <span class="p">(</span><span class="n">l2norm</span><span class="p">(</span><span class="n">s_zm</span><span class="p">,</span> <span class="n">keepdim</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span> <span class="o">**</span> <span class="mi">2</span> <span class="o">+</span> <span class="n">eps</span><span class="p">)</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="mi">20</span> <span class="o">*</span> <span class="n">torch</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">eps</span> <span class="o">+</span> <span class="n">l2norm</span><span class="p">(</span><span class="n">t</span><span class="p">)</span> <span class="o">/</span> <span class="p">(</span><span class="n">l2norm</span><span class="p">(</span><span class="n">x_zm</span> <span class="o">-</span> <span class="n">t</span><span class="p">)</span> <span class="o">+</span> <span class="n">eps</span><span class="p">))</span></div>


<div class="viewcode-block" id="SISNRLoss.forward">
<a class="viewcode-back" href="../../../clarity.engine.html#clarity.engine.losses.SISNRLoss.forward">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">forward</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">):</span>
        <span class="k">return</span> <span class="o">-</span><span class="bp">self</span><span class="o">.</span><span class="n">cal_sisnr</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">)</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span></div>
</div>



<div class="viewcode-block" id="SNRLoss">
<a class="viewcode-back" href="../../../clarity.engine.html#clarity.engine.losses.SNRLoss">[docs]</a>
<span class="k">class</span><span class="w"> </span><span class="nc">SNRLoss</span><span class="p">(</span><span class="n">torch</span><span class="o">.</span><span class="n">nn</span><span class="o">.</span><span class="n">Module</span><span class="p">):</span>
    <span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">tao</span><span class="o">=</span><span class="mf">1e-3</span><span class="p">):</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">tao</span> <span class="o">=</span> <span class="n">tao</span>

<div class="viewcode-block" id="SNRLoss.l2norm">
<a class="viewcode-back" href="../../../clarity.engine.html#clarity.engine.losses.SNRLoss.l2norm">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">l2norm</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">mat</span><span class="p">,</span> <span class="n">keepdim</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
        <span class="k">return</span> <span class="n">torch</span><span class="o">.</span><span class="n">norm</span><span class="p">(</span><span class="n">mat</span><span class="p">,</span> <span class="n">dim</span><span class="o">=-</span><span class="mi">1</span><span class="p">,</span> <span class="n">keepdim</span><span class="o">=</span><span class="n">keepdim</span><span class="p">)</span></div>


<div class="viewcode-block" id="SNRLoss.forward">
<a class="viewcode-back" href="../../../clarity.engine.html#clarity.engine.losses.SNRLoss.forward">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">forward</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">x</span><span class="p">,</span> <span class="n">s</span><span class="p">,</span> <span class="n">eps</span><span class="o">=</span><span class="mf">1e-8</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">x</span><span class="o">.</span><span class="n">shape</span> <span class="o">!=</span> <span class="n">s</span><span class="o">.</span><span class="n">shape</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">RuntimeError</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Dimension mismatch when calculate si-snr, </span><span class="si">{</span><span class="n">x</span><span class="o">.</span><span class="n">shape</span><span class="si">}</span><span class="s2"> vs </span><span class="si">{</span><span class="n">s</span><span class="o">.</span><span class="n">shape</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="p">)</span>

        <span class="n">loss</span> <span class="o">=</span> <span class="mi">10</span> <span class="o">*</span> <span class="n">torch</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">l2norm</span><span class="p">(</span><span class="n">s</span> <span class="o">-</span> <span class="n">x</span><span class="p">)</span> <span class="o">**</span> <span class="mi">2</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">tao</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">l2norm</span><span class="p">(</span><span class="n">s</span><span class="p">)</span> <span class="o">**</span> <span class="mi">2</span> <span class="o">+</span> <span class="n">eps</span>
        <span class="p">)</span> <span class="o">-</span> <span class="mi">10</span> <span class="o">*</span> <span class="n">torch</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">l2norm</span><span class="p">(</span><span class="n">s</span><span class="p">)</span> <span class="o">**</span> <span class="mi">2</span> <span class="o">+</span> <span class="n">eps</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">loss</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span></div>
</div>



<div class="viewcode-block" id="STOILoss">
<a class="viewcode-back" href="../../../clarity.engine.html#clarity.engine.losses.STOILoss">[docs]</a>
<span class="k">class</span><span class="w"> </span><span class="nc">STOILoss</span><span class="p">(</span><span class="n">torch</span><span class="o">.</span><span class="n">nn</span><span class="o">.</span><span class="n">Module</span><span class="p">):</span>
    <span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">sr</span><span class="p">):</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">NegSTOI</span> <span class="o">=</span> <span class="n">NegSTOILoss</span><span class="p">(</span><span class="n">sample_rate</span><span class="o">=</span><span class="n">sr</span><span class="p">)</span>

<div class="viewcode-block" id="STOILoss.forward">
<a class="viewcode-back" href="../../../clarity.engine.html#clarity.engine.losses.STOILoss.forward">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">forward</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">x</span><span class="p">,</span> <span class="n">s</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">NegSTOI</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">s</span><span class="p">)</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span></div>
</div>



<div class="viewcode-block" id="STOILevelLoss">
<a class="viewcode-back" href="../../../clarity.engine.html#clarity.engine.losses.STOILevelLoss">[docs]</a>
<span class="k">class</span><span class="w"> </span><span class="nc">STOILevelLoss</span><span class="p">(</span><span class="n">torch</span><span class="o">.</span><span class="n">nn</span><span class="o">.</span><span class="n">Module</span><span class="p">):</span>
    <span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">sr</span><span class="p">,</span> <span class="n">alpha</span><span class="p">,</span> <span class="n">block_size</span><span class="o">=</span><span class="mf">0.4</span><span class="p">,</span> <span class="n">overlap</span><span class="o">=</span><span class="mf">0.7</span><span class="p">,</span> <span class="n">gamma_a</span><span class="o">=-</span><span class="mi">70</span><span class="p">):</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">NegSTOI</span> <span class="o">=</span> <span class="n">NegSTOILoss</span><span class="p">(</span><span class="n">sample_rate</span><span class="o">=</span><span class="n">sr</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">alpha</span> <span class="o">=</span> <span class="n">alpha</span>

        <span class="s2">&quot;rms measurement&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">frame_size</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">block_size</span> <span class="o">*</span> <span class="n">sr</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">frame_shift</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">block_size</span> <span class="o">*</span> <span class="n">sr</span> <span class="o">*</span> <span class="p">(</span><span class="mi">1</span> <span class="o">-</span> <span class="n">overlap</span><span class="p">))</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">unfold</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">nn</span><span class="o">.</span><span class="n">Unfold</span><span class="p">(</span>
            <span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">frame_size</span><span class="p">),</span> <span class="n">stride</span><span class="o">=</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">frame_shift</span><span class="p">)</span>
        <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">gamma_a</span> <span class="o">=</span> <span class="n">gamma_a</span>

        <span class="s2">&quot;mse&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">cal_mse</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">nn</span><span class="o">.</span><span class="n">MSELoss</span><span class="p">()</span>

<div class="viewcode-block" id="STOILevelLoss.measure_loudness">
<a class="viewcode-back" href="../../../clarity.engine.html#clarity.engine.losses.STOILevelLoss.measure_loudness">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">measure_loudness</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">signal</span><span class="p">,</span> <span class="n">eps</span><span class="o">=</span><span class="mf">1e-8</span><span class="p">):</span>
        <span class="n">x_unfold</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">unfold</span><span class="p">(</span><span class="n">signal</span><span class="o">.</span><span class="n">unsqueeze</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">unsqueeze</span><span class="p">(</span><span class="mi">2</span><span class="p">))</span>

        <span class="n">z</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">torch</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">x_unfold</span><span class="o">**</span><span class="mi">2</span><span class="p">,</span> <span class="n">dim</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span> <span class="o">/</span> <span class="bp">self</span><span class="o">.</span><span class="n">frame_size</span>
        <span class="p">)</span>  <span class="c1"># mean square for each frame</span>
        <span class="n">el</span> <span class="o">=</span> <span class="o">-</span><span class="mf">0.691</span> <span class="o">+</span> <span class="mi">10</span> <span class="o">*</span> <span class="n">torch</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">z</span> <span class="o">+</span> <span class="n">eps</span><span class="p">)</span>

        <span class="n">idx_a</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">el</span> <span class="o">&gt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">gamma_a</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
        <span class="n">z_ave_gated_a</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">z</span> <span class="o">*</span> <span class="n">idx_a</span><span class="p">,</span> <span class="n">dim</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">keepdim</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span> <span class="o">/</span> <span class="p">(</span>
            <span class="n">torch</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">idx_a</span><span class="p">,</span> <span class="n">dim</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">keepdim</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span> <span class="o">+</span> <span class="n">eps</span>
        <span class="p">)</span>
        <span class="n">gamma_r</span> <span class="o">=</span> <span class="o">-</span><span class="mf">0.691</span> <span class="o">+</span> <span class="mi">10</span> <span class="o">*</span> <span class="n">torch</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">z_ave_gated_a</span> <span class="o">+</span> <span class="n">eps</span><span class="p">)</span> <span class="o">-</span> <span class="mi">10</span>

        <span class="n">idx_r</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">el</span> <span class="o">&gt;</span> <span class="n">gamma_r</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
        <span class="n">idx_a_r</span> <span class="o">=</span> <span class="n">idx_a</span> <span class="o">*</span> <span class="n">idx_r</span>
        <span class="n">z_ave_gated_a_r</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">z</span> <span class="o">*</span> <span class="n">idx_a_r</span><span class="p">,</span> <span class="n">dim</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">keepdim</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span> <span class="o">/</span> <span class="p">(</span>
            <span class="n">torch</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">idx_a_r</span><span class="p">,</span> <span class="n">dim</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">keepdim</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span> <span class="o">+</span> <span class="n">eps</span>
        <span class="p">)</span>
        <span class="n">lufs</span> <span class="o">=</span> <span class="o">-</span><span class="mf">0.691</span> <span class="o">+</span> <span class="mi">10</span> <span class="o">*</span> <span class="n">torch</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="n">z_ave_gated_a_r</span> <span class="o">+</span> <span class="n">eps</span><span class="p">)</span>  <span class="c1"># loudness</span>
        <span class="k">return</span> <span class="n">lufs</span></div>


<div class="viewcode-block" id="STOILevelLoss.forward">
<a class="viewcode-back" href="../../../clarity.engine.html#clarity.engine.losses.STOILevelLoss.forward">[docs]</a>
    <span class="k">def</span><span class="w"> </span><span class="nf">forward</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">x</span><span class="p">,</span> <span class="n">s</span><span class="p">):</span>
        <span class="n">loudness_x</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">measure_loudness</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>
        <span class="n">loudness_s</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">measure_loudness</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>
        <span class="n">LevelLoss</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">alpha</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">cal_mse</span><span class="p">(</span><span class="n">loudness_x</span><span class="p">,</span> <span class="n">loudness_s</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">LevelLoss</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">NegSTOI</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">s</span><span class="p">)</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span></div>
</div>

</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2020-2025, pyClarity authors.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>