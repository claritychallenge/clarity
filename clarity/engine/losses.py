import torch

from clarity.predictor.torch_stoi import NegSTOILoss


class SISNRLoss(torch.nn.Module):
    # def __init__(self):  # removed as does nothing
    #    super(SISNRLoss, self).__init__()

    def cal_sisnr(self, x, s, eps=1e-8):
        """
        Arguments:
        x: separated signal, N x S tensor
        s: reference signal, N x S tensor
        Return:
        sisnr: N tensor
        """

        def l2norm(mat, keepdim=False):
            return torch.norm(mat, dim=-1, keepdim=keepdim)

        if x.shape != s.shape:
            raise RuntimeError(
                f"Dimension mismatch when calculate si-snr, {x.shape} vs {s.shape}"
            )
        x_zm = x - torch.mean(x, dim=-1, keepdim=True)
        s_zm = s - torch.mean(s, dim=-1, keepdim=True)
        t = (
            torch.sum(x_zm * s_zm, dim=-1, keepdim=True)
            * s_zm
            / (l2norm(s_zm, keepdim=True) ** 2 + eps)
        )
        return 20 * torch.log10(eps + l2norm(t) / (l2norm(x_zm - t) + eps))

    def forward(self, x, y):
        return -self.cal_sisnr(x, y).mean()


class SNRLoss(torch.nn.Module):
    def __init__(self, tao=1e-3):
        super().__init__()
        self.tao = tao

    def l2norm(self, mat, keepdim=False):
        return torch.norm(mat, dim=-1, keepdim=keepdim)

    def forward(self, x, s, eps=1e-8):
        if x.shape != s.shape:
            raise RuntimeError(
                f"Dimension mismatch when calculate si-snr, {x.shape} vs {s.shape}"
            )

        loss = 10 * torch.log10(
            self.l2norm(s - x) ** 2 + self.tao * self.l2norm(s) ** 2 + eps
        ) - 10 * torch.log10(self.l2norm(s) ** 2 + eps)
        return loss.mean()


class STOILoss(torch.nn.Module):
    def __init__(self, sr):
        super().__init__()
        self.NegSTOI = NegSTOILoss(sample_rate=sr)

    def forward(self, x, s):
        return self.NegSTOI(x, s).mean()


class STOILevelLoss(torch.nn.Module):
    def __init__(self, sr, alpha, block_size=0.4, overlap=0.7, gamma_a=-70):
        super().__init__()
        self.NegSTOI = NegSTOILoss(sample_rate=sr)
        self.alpha = alpha

        "rms measurement"
        self.frame_size = int(block_size * sr)
        self.frame_shift = int(block_size * sr * (1 - overlap))
        self.unfold = torch.nn.Unfold(
            (1, self.frame_size), stride=(1, self.frame_shift)
        )
        self.gamma_a = gamma_a

        "mse"
        self.cal_mse = torch.nn.MSELoss()

    def measure_loudness(self, signal, eps=1e-8):
        x_unfold = self.unfold(signal.unsqueeze(1).unsqueeze(2))

        z = (
            torch.sum(x_unfold**2, dim=1) / self.frame_size
        )  # mean square for each frame
        el = -0.691 + 10 * torch.log10(z + eps)

        idx_a = torch.where(el > self.gamma_a, 1, 0)
        z_ave_gated_a = torch.sum(z * idx_a, dim=1, keepdim=True) / (
            torch.sum(idx_a, dim=1, keepdim=True) + eps
        )
        gamma_r = -0.691 + 10 * torch.log10(z_ave_gated_a + eps) - 10

        idx_r = torch.where(el > gamma_r, 1, 0)
        idx_a_r = idx_a * idx_r
        z_ave_gated_a_r = torch.sum(z * idx_a_r, dim=1, keepdim=True) / (
            torch.sum(idx_a_r, dim=1, keepdim=True) + eps
        )
        lufs = -0.691 + 10 * torch.log10(z_ave_gated_a_r + eps)  # loudness
        return lufs

    def forward(self, x, s):
        loudness_x = self.measure_loudness(x)
        loudness_s = self.measure_loudness(s)
        LevelLoss = self.alpha * self.cal_mse(loudness_x, loudness_s)
        return LevelLoss + self.NegSTOI(x, s).mean()
