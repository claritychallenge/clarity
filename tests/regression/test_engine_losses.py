# Regression test

import torch

from clarity.engine.losses import SISNRLoss, SNRLoss, STOILevelLoss, STOILoss


def test_sisnr_loss(regtest):
    torch.manual_seed(0)
    torch.set_num_threads(1)
    si_snr_loss = SISNRLoss()
    x = torch.randn(10, 1000)
    y = torch.randn(10, 1000)
    loss = si_snr_loss.forward(x, y)

    regtest.write(f"SISNR loss {loss:0.4f}\n")


def test_snr_loss(regtest):
    torch.manual_seed(0)
    torch.set_num_threads(1)
    snr_loss = SNRLoss()
    x = torch.randn(10, 1000)
    y = torch.randn(10, 1000)
    loss = snr_loss.forward(x, y)

    regtest.write(f"SNR loss {loss:0.5f}\n")


def test_stoi_loss(regtest):
    torch.manual_seed(0)
    torch.set_num_threads(1)
    stoi_loss = STOILoss(sr=16000)
    x = torch.randn(2, 16000)
    y = torch.randn(2, 16000)
    loss = stoi_loss.forward(x, y)

    regtest.write(f"STOI loss {loss:0.7f}\n")


def test_stoi_level_loss(regtest):
    torch.manual_seed(0)
    torch.set_num_threads(1)
    stoi_level_loss = STOILevelLoss(sr=16000, alpha=0.5)
    x = torch.randn(2, 16000)
    y = torch.randn(2, 16000)
    loss = stoi_level_loss.forward(x, y)

    regtest.write(f"STOI level loss {loss:0.7f}\n")
