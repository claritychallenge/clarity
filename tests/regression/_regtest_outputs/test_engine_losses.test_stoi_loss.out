STOI loss 0.0164208
